## Exploring London Wards

the City of London is one of the most important and international cities. It is know with many nouns, one is "The City of London" or simply "the City". It covers a total area of 1569 km^2 and it is divided in 25 Wards in Total.

This analysis is addressed to those who seek to open a Pizza place in London. The goal is to understand which area is the most suitable.

I found on that link: https://www.doogal.co.uk/AdministrativeAreas.php?district=E09000001 all the Wards associated to latitude and longitude in London. It is a csv file that and I will use it later. In this csv the information that I gained are: the postal codes, latitude, longitude, wards and some other data that, honestly, are not relevant for my analysis. 

I will use Foursquare location data to explore the Wards and see what types of restourant we can find in each of them. What I want to find with the use of the Foursquare location data is which ward has many pizza restaurants inside it, as well as other types of restaurant. 
Given the dataset with all the Ward I will explore each ward, look at the number of restaurants, look which ward is the suitable for opening a pizza place. 

My metric to decide if it is convinient or not to open a pizza place is based on the concentartion of competition.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import folium 

In [3]:
import numpy as np # library to handle data in a vectorized manner
import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import geopy
import geopandas

In [19]:
df=pd.read_csv('London_postcodes.csv')

In [20]:
df.shape

(6799, 16)

In [21]:
df.columns

Index(['Postcode', 'In Use?', 'Latitude', 'Longitude', 'Easting', 'Northing',
       'Grid Ref', 'Ward', 'Parish', 'Introduced', 'Terminated', 'Altitude',
       'Country', 'Last Updated', 'Quality', 'LSOA Code'],
      dtype='object')

In [22]:
df=df[df['In Use?']=='Yes']
df.shape

(1709, 16)

## The Data

As I said in the introduction, I found the data on link that follows: https://www.doogal.co.uk/AdministrativeAreas.php?district=E09000001. 
Here we can find all the Wards associated to latitude and longitude in London. The format of the data I get is a csv file. 

In this csv the information that I gained are: the postal codes, latitude, longitude, wards and some other data that, honestly, are not relevant for my analysis.

Thanks to this file I got the Wards and their geographical coordinates, but there are many different latitudes for each ward because of the different postal codes in each area. For simplicity I will use just one of them.

The dataset consists in 6799 observation, but the number is consistently reduced by selecting only the existing postal codes, that still in use. 

In this notebook I selected only the Wards and their geographical coordinates, and I dropped all the other columns. Later on I will use the Foursquare API to explore each ward and look at the piazza restaurants inside each one of them.

In [23]:
df.head()

Postcode In Use?   Latitude  Longitude  Easting  Northing  Grid Ref  \
0   E1 6AN     Yes  51.518895  -0.078378   533425    181747  TQ334817   
1   E1 7AA     Yes  51.515567  -0.075635   533625    181382  TQ336813   
2   E1 7AD     Yes  51.515457  -0.076718   533550    181368  TQ335813   
3   E1 7AE     Yes  51.515613  -0.076899   533537    181385  TQ335813   
4   E1 7AF     Yes  51.515613  -0.076899   533537    181385  TQ335813   

          Ward                           Parish  Introduced Terminated  \
0  Bishopsgate  City of London, unparished area  1980-01-01        NaN   
1    Portsoken  City of London, unparished area  2000-12-01        NaN   
2    Portsoken  City of London, unparished area  2013-09-01        NaN   
3    Portsoken  City of London, unparished area  2013-07-01        NaN   
4    Portsoken  City of London, unparished area  2013-01-01        NaN   

   Altitude  Country Last Updated  \
0        32  England   2020-02-19   
1        28  England   2020-02-19   
2        31  England   2020-02-19   
3        30  England   2020-02-19   
4        30  England   2020-02-19   

                                             Quality  LSOA Code  
0  Within the building of the matched address clo...  E01032739  
1  Within the building of the matched address clo...  E01000005  
2  Within the building of the matched address clo...  E01000005  
3  Within the building of the matched address clo...  E01000005  
4  Within the building of the matched address clo...  E01000005

In [24]:
df1=df.sort_values(by="Last Updated").drop_duplicates(subset=["Ward"], keep="last")

In [25]:
df1.reset_index(inplace=True)
df1.head()

index  Postcode In Use?   Latitude  Longitude  Easting  Northing  Grid Ref  \
0   5878  EC4V 2AD     Yes  51.511636  -0.094937   532297    180910  TQ322809   
1   5877  EC4V 2AB     Yes  51.512132  -0.093893   532368    180967  TQ323809   
2   5805  EC4R 9AF     Yes  51.510483  -0.086540   532883    180797  TQ328807   
3   5802  EC4R 9AB     Yes  51.510225  -0.088338   532759    180765  TQ327807   
4    482  EC1N 2HA     Yes  51.517092  -0.107537   531407    181494  TQ314814   

             Ward                           Parish  Introduced Terminated  \
0      Queenhithe  City of London, unparished area  2009-07-01        NaN   
1          Vintry  City of London, unparished area  1991-09-01        NaN   
2      Candlewick  City of London, unparished area  1999-12-01        NaN   
3         Dowgate  City of London, unparished area  1980-01-01        NaN   
4  Castle Baynard  City of London, unparished area  1998-12-01        NaN   

   Altitude  Country Last Updated  \
0        22  England   2020-02-19   
1        27  England   2020-02-19   
2        31  England   2020-02-19   
3        25  England   2020-02-19   
4        31  England   2020-02-19   

                                             Quality  LSOA Code  
0  Within the building of the matched address clo...  E01032739  
1  Within the building of the matched address clo...  E01032739  
2  Within the building of the matched address clo...  E01032739  
3  Within the building of the matched address clo...  E01032739  
4  Within the building of the matched address clo...  E01032740

In [26]:
len(df1)

25

In [107]:
#list of all Wards
df1.Ward.value_counts()

Bread Street          1
Cordwainer            1
Farringdon Without    1
Cripplegate           1
Candlewick            1
Dowgate               1
Portsoken             1
Langbourn             1
Walbrook              1
Billingsgate          1
Vintry                1
Bridge                1
Bishopsgate           1
Farringdon Within     1
Broad Street          1
Bassishaw             1
Cornhill              1
Castle Baynard        1
Aldersgate            1
Tower                 1
Cheap                 1
Queenhithe            1
Aldgate               1
Coleman Street        1
Lime Street           1
Name: Ward, dtype: int64

In [28]:
df1.columns

Index(['index', 'Postcode', 'In Use?', 'Latitude', 'Longitude', 'Easting',
       'Northing', 'Grid Ref', 'Ward', 'Parish', 'Introduced', 'Terminated',
       'Altitude', 'Country', 'Last Updated', 'Quality', 'LSOA Code'],
      dtype='object')

## Creation of the dataset with only Wards and their latitude and longitude


In [29]:
df2=df1[['Ward','Latitude', 'Longitude']]
df2.head()

Ward   Latitude  Longitude
0      Queenhithe  51.511636  -0.094937
1          Vintry  51.512132  -0.093893
2      Candlewick  51.510483  -0.086540
3         Dowgate  51.510225  -0.088338
4  Castle Baynard  51.517092  -0.107537

In [71]:
#plot of the different wards
Latitude = location.latitude
Longitude = location.longitude
# create map of Manhattan using latitude and longitude values
map = folium.Map(location=[Latitude, Longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(df2['Latitude'], df2['Longitude'], df2['Ward']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map)  
    
map

## London City venues

In [32]:
CLIENT_ID = 'DV5DFTF4YGRFQD51YWF3OP10AEYMOJJX1MGWV0MLBEQUK0MC' # your Foursquare ID
CLIENT_SECRET = 'XJQ0Z2BDLT4Y2N1SRHYLFDKDRILBV3RRWMO5HWKTZG2RLXKF' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DV5DFTF4YGRFQD51YWF3OP10AEYMOJJX1MGWV0MLBEQUK0MC
CLIENT_SECRET:XJQ0Z2BDLT4Y2N1SRHYLFDKDRILBV3RRWMO5HWKTZG2RLXKF


In [126]:
LIMIT = 300 # limit of number of venues returned by Foursquare API
radius = 1000 # define radius

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Ward', 
                  'Ward Latitude', 
                  'Ward Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [127]:
# type your answer here

London_venues = getNearbyVenues(names=df2['Ward'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )

Queenhithe
Vintry
Candlewick
Dowgate
Castle Baynard
Farringdon Within
Bishopsgate
Portsoken
Lime Street
Langbourn
Aldgate
Billingsgate
Bridge
Tower
Bread Street
Cornhill
Broad Street
Cripplegate
Walbrook
Cordwainer
Coleman Street
Aldersgate
Bassishaw
Cheap
Farringdon Without


In [128]:
print(London_venues.shape)
London_venues

(2392, 7)


Ward  Ward Latitude  Ward Longitude                Venue  \
0             Queenhithe      51.511636       -0.094937              Rosslyn   
1             Queenhithe      51.511636       -0.094937  Porterford Butchers   
2             Queenhithe      51.511636       -0.094937   The Merchant House   
3             Queenhithe      51.511636       -0.094937                 Host   
4             Queenhithe      51.511636       -0.094937      M&S Simply Food   
...                  ...            ...             ...                  ...   
2387  Farringdon Without      51.517484       -0.112941          Soho Coffee   
2388  Farringdon Without      51.517484       -0.112941     Vital Ingredient   
2389  Farringdon Without      51.517484       -0.112941              El Vino   
2390  Farringdon Without      51.517484       -0.112941        Pret A Manger   
2391  Farringdon Without      51.517484       -0.112941      Daly's Wine Bar   

      Venue Latitude  Venue Longitude  Venue Category  
0          51.512574        -0.093381     Coffee Shop  
1          51.513032        -0.093737         Butcher  
2          51.513264        -0.093039    Cocktail Bar  
3          51.512629        -0.093211     Coffee Shop  
4          51.513590        -0.095297   Grocery Store  
...              ...              ...             ...  
2387       51.513333        -0.113325     Coffee Shop  
2388       51.514238        -0.109501     Salad Place  
2389       51.514175        -0.109422        Wine Bar  
2390       51.513432        -0.112947  Sandwich Place  
2391       51.513139        -0.113428        Wine Bar  

[2392 rows x 7 columns]

In [129]:
London_venues['Venue Category'].unique()

array(['Coffee Shop', 'Butcher', 'Cocktail Bar', 'Grocery Store', 'Café',
       'Roof Deck', 'Vietnamese Restaurant', 'Gym / Fitness Center',
       'Wine Bar', 'Seafood Restaurant', 'Hotel', 'Italian Restaurant',
       'Bookstore', 'Udon Restaurant', 'Chinese Restaurant', 'Park',
       'Sandwich Place', 'Restaurant', 'Indian Restaurant',
       'Scandinavian Restaurant', 'Bar', 'Modern European Restaurant',
       'English Restaurant', 'Pizza Place', 'French Restaurant',
       'Theater', 'Asian Restaurant', 'Pedestrian Plaza', 'Gym', 'Lounge',
       'Scenic Lookout', 'Steakhouse', 'Burger Joint', 'Food Truck',
       'Mexican Restaurant', 'History Museum', 'Falafel Restaurant',
       'Nightclub', 'New American Restaurant', 'Bakery', 'Shopping Mall',
       'Deli / Bodega', 'Clothing Store', 'Plaza', 'American Restaurant',
       'Fast Food Restaurant', 'Art Gallery', 'Hotel Bar',
       'Sushi Restaurant', 'Juice Bar', 'Trail', 'Pub',
       'Turkish Restaurant', 'Salad Place', 

In [130]:
pizza_places=London_venues[London_venues['Venue Category']=='Pizza Place']
pizza_places

Ward  Ward Latitude  Ward Longitude              Venue  \
28           Queenhithe      51.511636       -0.094937          Homeslice   
76           Queenhithe      51.511636       -0.094937       Franco Manca   
132              Vintry      51.512132       -0.093893          Homeslice   
184              Vintry      51.512132       -0.093893       Franco Manca   
276          Candlewick      51.510483       -0.086540          Homeslice   
349             Dowgate      51.510225       -0.088338          Homeslice   
481   Farringdon Within      51.519765       -0.099199              Pixxa   
552         Bishopsgate      51.515796       -0.080380        Pizza Union   
567         Bishopsgate      51.515796       -0.080380  1n1 Fashion Pizza   
591         Bishopsgate      51.515796       -0.080380       Franco Manca   
641           Portsoken      51.513976       -0.076305  1n1 Fashion Pizza   
668           Portsoken      51.513976       -0.076305        Pizza Union   
672           Portsoken      51.513976       -0.076305        Pizza Union   
759         Lime Street      51.515317       -0.080241        Pizza Union   
762         Lime Street      51.515317       -0.080241  1n1 Fashion Pizza   
808         Lime Street      51.515317       -0.080241       Franco Manca   
1009            Aldgate      51.512236       -0.079347  1n1 Fashion Pizza   
1245              Tower      51.512195       -0.074680        Pizza Union   
1270              Tower      51.512195       -0.074680  1n1 Fashion Pizza   
1343       Bread Street      51.514751       -0.097401       Franco Manca   
1400       Bread Street      51.514751       -0.097401          Homeslice   
1567       Broad Street      51.515891       -0.087729       Franco Manca   
1761           Walbrook      51.514843       -0.090756          Homeslice   
1831         Cordwainer      51.513583       -0.092997          Homeslice   
1962     Coleman Street      51.515079       -0.090974          Homeslice   
2053         Aldersgate      51.516735       -0.096065       Franco Manca   
2191          Bassishaw      51.516650       -0.093027          Homeslice   
2252              Cheap      51.514724       -0.094465          Homeslice   
2279              Cheap      51.514724       -0.094465       Franco Manca   

      Venue Latitude  Venue Longitude Venue Category  
28         51.512259        -0.092557    Pizza Place  
76         51.513731        -0.100302    Pizza Place  
132        51.512259        -0.092557    Pizza Place  
184        51.513731        -0.100302    Pizza Place  
276        51.512259        -0.092557    Pizza Place  
349        51.512259        -0.092557    Pizza Place  
481        51.520053        -0.101632    Pizza Place  
552        51.517699        -0.077416    Pizza Place  
567        51.516037        -0.075865    Pizza Place  
591        51.518780        -0.083448    Pizza Place  
641        51.516037        -0.075865    Pizza Place  
668        51.517699        -0.077416    Pizza Place  
672        51.513823        -0.070529    Pizza Place  
759        51.517699        -0.077416    Pizza Place  
762        51.516037        -0.075865    Pizza Place  
808        51.518780        -0.083448    Pizza Place  
1009       51.516037        -0.075865    Pizza Place  
1245       51.513823        -0.070529    Pizza Place  
1270       51.516037        -0.075865    Pizza Place  
1343       51.513731        -0.100302    Pizza Place  
1400       51.512259        -0.092557    Pizza Place  
1567       51.518780        -0.083448    Pizza Place  
1761       51.512259        -0.092557    Pizza Place  
1831       51.512259        -0.092557    Pizza Place  
1962       51.512259        -0.092557    Pizza Place  
2053       51.513731        -0.100302    Pizza Place  
2191       51.512259        -0.092557    Pizza Place  
2252       51.512259        -0.092557    Pizza Place  
2279       51.513731        -0.100302    Pizza Place

In [131]:
pizza_places.Ward.value_counts()

Lime Street          3
Portsoken            3
Bishopsgate          3
Cheap                2
Vintry               2
Bread Street         2
Tower                2
Queenhithe           2
Bassishaw            1
Cordwainer           1
Farringdon Within    1
Aldersgate           1
Broad Street         1
Walbrook             1
Coleman Street       1
Aldgate              1
Dowgate              1
Candlewick           1
Name: Ward, dtype: int64

## Firstly Analyse each Ward

In [132]:
# one hot encoding
London_onehot = pd.get_dummies(London_venues[['Venue Category']], prefix="", prefix_sep="")
# add neighborhood column back to dataframe
London_onehot['Ward'] = London_venues['Ward'] 

# move neighborhood column to the first column
fixed_columns = [London_onehot.columns[-1]] + list(London_onehot.columns[:-1])
London_onehot = London_onehot[fixed_columns]
London_onehot.head()

Ward  American Restaurant  Arcade  Argentinian Restaurant  \
0  Queenhithe                    0       0                       0   
1  Queenhithe                    0       0                       0   
2  Queenhithe                    0       0                       0   
3  Queenhithe                    0       0                       0   
4  Queenhithe                    0       0                       0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  BBQ Joint  Bagel Shop  \
0            0                    0                 0          0           0   
1            0                    0                 0          0           0   
2            0                    0                 0          0           0   
3            0                    0                 0          0           0   
4            0                    0                 0          0           0   

   Bakery  ...  Trail  Turkish Restaurant  Udon Restaurant  \
0       0  ...      0                   0                0   
1       0  ...      0                   0                0   
2       0  ...      0                   0                0   
3       0  ...      0                   0                0   
4       0  ...      0                   0                0   

   Vegetarian / Vegan Restaurant  Vietnamese Restaurant  Whisky Bar  Wine Bar  \
0                              0                      0           0         0   
1                              0                      0           0         0   
2                              0                      0           0         0   
3                              0                      0           0         0   
4                              0                      0           0         0   

   Wine Shop  Women's Store  Yoga Studio  
0          0              0            0  
1          0              0            0  
2          0              0            0  
3          0              0            0  
4          0              0            0  

[5 rows x 142 columns]

In [133]:
London_grouped = London_onehot.groupby('Ward').mean().reset_index()
London_grouped

Ward  American Restaurant  Arcade  Argentinian Restaurant  \
0           Aldersgate                 0.00    0.00                0.000000   
1              Aldgate                 0.00    0.00                0.020000   
2            Bassishaw                 0.00    0.00                0.000000   
3         Billingsgate                 0.00    0.00                0.010526   
4          Bishopsgate                 0.00    0.00                0.010000   
5         Bread Street                 0.01    0.00                0.000000   
6               Bridge                 0.00    0.00                0.010000   
7         Broad Street                 0.00    0.00                0.010000   
8           Candlewick                 0.00    0.00                0.011628   
9       Castle Baynard                 0.00    0.00                0.010000   
10               Cheap                 0.00    0.00                0.000000   
11      Coleman Street                 0.00    0.00                0.000000   
12          Cordwainer                 0.00    0.00                0.000000   
13            Cornhill                 0.00    0.00                0.010000   
14         Cripplegate                 0.00    0.00                0.000000   
15             Dowgate                 0.00    0.00                0.012048   
16   Farringdon Within                 0.00    0.00                0.000000   
17  Farringdon Without                 0.00    0.01                0.010000   
18           Langbourn                 0.00    0.00                0.010000   
19         Lime Street                 0.00    0.00                0.020000   
20           Portsoken                 0.00    0.00                0.010000   
21          Queenhithe                 0.01    0.00                0.000000   
22               Tower                 0.00    0.00                0.011765   
23              Vintry                 0.00    0.00                0.000000   
24            Walbrook                 0.00    0.00                0.000000   

    Art Gallery  Arts & Crafts Store  Asian Restaurant  BBQ Joint  Bagel Shop  \
0      0.030000             0.000000          0.000000   0.000000    0.000000   
1      0.000000             0.000000          0.020000   0.010000    0.000000   
2      0.030000             0.000000          0.010000   0.000000    0.000000   
3      0.000000             0.000000          0.031579   0.010526    0.000000   
4      0.000000             0.000000          0.010000   0.000000    0.000000   
5      0.010000             0.000000          0.010000   0.000000    0.000000   
6      0.000000             0.000000          0.030000   0.010000    0.000000   
7      0.010000             0.000000          0.020000   0.000000    0.000000   
8      0.000000             0.000000          0.046512   0.000000    0.000000   
9      0.000000             0.000000          0.010000   0.000000    0.000000   
10     0.010000             0.000000          0.020000   0.000000    0.000000   
11     0.010000             0.000000          0.020000   0.000000    0.000000   
12     0.010000             0.000000          0.030000   0.000000    0.000000   
13     0.010000             0.000000          0.030000   0.000000    0.000000   
14     0.039474             0.000000          0.000000   0.000000    0.000000   
15     0.000000             0.000000          0.036145   0.000000    0.000000   
16     0.029851             0.000000          0.000000   0.000000    0.014925   
17     0.000000             0.000000          0.010000   0.010000    0.000000   
18     0.000000             0.000000          0.020000   0.010000    0.000000   
19     0.000000             0.000000          0.010000   0.000000    0.000000   
20     0.010000             0.000000          0.020000   0.010000    0.000000   
21     0.010000             0.000000          0.030000   0.000000    0.000000   
22     0.000000             0.011765          0.023529   0.011765    0.000000   
23     0.010000    

In [134]:
num_top_venues = 15
for hood in London_grouped['Ward']:
    print("----"+hood+"----")
    temp = London_grouped[London_grouped['Ward'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Aldersgate----
                         venue  freq
0                  Coffee Shop  0.08
1           Italian Restaurant  0.05
2             Sushi Restaurant  0.05
3   Modern European Restaurant  0.04
4                        Plaza  0.04
5         Gym / Fitness Center  0.04
6                  Art Gallery  0.03
7               Clothing Store  0.03
8                 Cocktail Bar  0.03
9                   Steakhouse  0.03
10                Burger Joint  0.03
11                  Restaurant  0.03
12                Concert Hall  0.02
13                        Park  0.02
14                        Café  0.02


----Aldgate----
                   venue  freq
0                  Hotel  0.11
1   Gym / Fitness Center  0.07
2           Cocktail Bar  0.05
3             Restaurant  0.04
4            Coffee Shop  0.04
5                    Pub  0.03
6     English Restaurant  0.03
7                 Garden  0.03
8      French Restaurant  0.03
9       Asian Restaurant  0.02
10             Hotel Bar  0.02

In [135]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [136]:
num_top_venues = 15
indicators = ['st', 'nd', 'rd']
# create columns according to number of top venues
columns = ['Ward']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
Ward_venues_sorted = pd.DataFrame(columns=columns)
Ward_venues_sorted['Ward'] = London_grouped['Ward']

for ind in np.arange(London_grouped.shape[0]):
    Ward_venues_sorted.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

Ward_venues_sorted.head()

Ward 1st Most Common Venue 2nd Most Common Venue  \
0    Aldersgate           Coffee Shop    Italian Restaurant   
1       Aldgate                 Hotel  Gym / Fitness Center   
2     Bassishaw           Coffee Shop    Italian Restaurant   
3  Billingsgate                 Hotel            Restaurant   
4   Bishopsgate           Coffee Shop          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0      Sushi Restaurant                 Plaza  Gym / Fitness Center   
1          Cocktail Bar            Restaurant           Coffee Shop   
2                  Café        Scenic Lookout           Art Gallery   
3           Coffee Shop  Gym / Fitness Center                   Pub   
4  Gym / Fitness Center     Indian Restaurant           Pizza Place   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Modern European Restaurant          Cocktail Bar          Burger Joint   
1           French Restaurant                Garden    English Restaurant   
2              Clothing Store            Steakhouse  Gym / Fitness Center   
3           French Restaurant                  Café          Cocktail Bar   
4    Mediterranean Restaurant           Salad Place                   Pub   

  9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
0            Restaurant             Steakhouse            Art Gallery   
1                   Pub            Salad Place                   Café   
2      Sushi Restaurant                 Lounge              Roof Deck   
3           Salad Place       Asian Restaurant                 Garden   
4                 Hotel             Restaurant     Italian Restaurant   

  12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
0         Clothing Store         Sandwich Place                 Garden   
1         Scenic Lookout              Hotel Bar      Indian Restaurant   
2             Restaurant                  Plaza                   Park   
3     Turkish Restaurant   Fast Food Restaurant           Burger Joint   
4    Japanese Restaurant             Boxing Gym         Sandwich Place   

      15th Most Common Venue  
0              Grocery Store  
1         Italian Restaurant  
2                Yoga Studio  
3                   Beer Bar  
4  Middle Eastern Restaurant

From this first analysis we could say that there is not a high precence of pizza places, but a high presence of italian restaurants. This is good for my puposes but I also have to admit that probabily Italian restaurants also do pizza, and to be more precise in my analysis I will focus also on this category. 

Another consideration that is important to do is that many of the pizza places are chain stores, and the place that I suggest to open is more likely a family conduction store, that is more the "Italian type". This leads to a completely different type of competitors.

## Clustering Wards

Since the London City Area is not so big I decided to cluster the venues in three clusters. 
Thanks to this clusterization I will explore each cluster and look at the one were the precence of Italian food is higher. I will not focus only on pizza places because there are just a few of them, and this is the reason why I decided to open one of them. I will also focus my attention on the Italian Restaurant, that probabily will do also pizza. 

My goal is to find a division in the clusters that allows me to say that one cluster is better Than the other.

In [137]:
# set number of clusters
kclusters = 3
London_clustering = London_grouped.drop('Ward', 1)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(London_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 0, 1, 1, 0, 1, 0, 0, 2])

In [138]:
# add clustering labels
Ward_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
London_merged = df2
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
London_merged = London_merged.join(Ward_venues_sorted.set_index('Ward'), on='Ward')

London_merged.head() # check the last columns!

Ward   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0      Queenhithe  51.511636  -0.094937               0           Coffee Shop   
1          Vintry  51.512132  -0.093893               0           Coffee Shop   
2      Candlewick  51.510483  -0.086540               0           Coffee Shop   
3         Dowgate  51.510225  -0.088338               0           Coffee Shop   
4  Castle Baynard  51.517092  -0.107537               2           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Italian Restaurant              Wine Bar  Gym / Fitness Center   
1    Italian Restaurant            Restaurant          Burger Joint   
2            Restaurant      Asian Restaurant                   Pub   
3                   Pub    Italian Restaurant            Restaurant   
4    Italian Restaurant                   Pub        Sandwich Place   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                    Pub    Seafood Restaurant   
1  Gym / Fitness Center  Vietnamese Restaurant      Asian Restaurant   
2  Gym / Fitness Center     Italian Restaurant                 Hotel   
3  Gym / Fitness Center          Historic Site                 Hotel   
4     French Restaurant   Gym / Fitness Center              Wine Bar   

        8th Most Common Venue       9th Most Common Venue  \
0            Asian Restaurant  Modern European Restaurant   
1  Modern European Restaurant            Sushi Restaurant   
2               Historic Site        Fast Food Restaurant   
3                  Steakhouse            Asian Restaurant   
4                       Hotel               Burrito Place   

  10th Most Common Venue 11th Most Common Venue 12th Most Common Venue  \
0           Burger Joint      French Restaurant  Vietnamese Restaurant   
1     Seafood Restaurant           Cocktail Bar                 Lounge   
2           Burger Joint                 Garden         Sandwich Place   
3           Burger Joint                   Café                  Trail   
4               Beer Bar   Fast Food Restaurant            Salad Place   

  13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  
0       Pedestrian Plaza         Scenic Lookout     Falafel Restaurant  
1            Salad Place             Steakhouse                    Bar  
2      Indian Restaurant      French Restaurant     Seafood Restaurant  
3     Seafood Restaurant           Cocktail Bar                Butcher  
4    Japanese Restaurant                    Bar     Falafel Restaurant

In [139]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1,2, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(London_merged['Latitude'], London_merged['Longitude'], London_merged['Ward'], London_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Exploring Clusters

Cluster 1

In [140]:
first=London_merged.loc[London_merged['Cluster Labels'] == 0, London_merged.columns[[0] + list(range(4, London_merged.shape[1]))]]
first

Ward 1st Most Common Venue 2nd Most Common Venue  \
0          Queenhithe           Coffee Shop    Italian Restaurant   
1              Vintry           Coffee Shop    Italian Restaurant   
2          Candlewick           Coffee Shop            Restaurant   
3             Dowgate           Coffee Shop                   Pub   
5   Farringdon Within                  Café                 Hotel   
14       Bread Street           Coffee Shop    Italian Restaurant   
15           Cornhill           Coffee Shop            Restaurant   
16       Broad Street           Coffee Shop                 Hotel   
17        Cripplegate           Coffee Shop  Gym / Fitness Center   
18           Walbrook           Coffee Shop    Italian Restaurant   
19         Cordwainer           Coffee Shop    Italian Restaurant   
20     Coleman Street           Coffee Shop    Italian Restaurant   
21         Aldersgate           Coffee Shop    Italian Restaurant   
22          Bassishaw           Coffee Shop    Italian Restaurant   
23              Cheap           Coffee Shop    Italian Restaurant   

   3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0               Wine Bar  Gym / Fitness Center             Restaurant   
1             Restaurant          Burger Joint   Gym / Fitness Center   
2       Asian Restaurant                   Pub   Gym / Fitness Center   
3     Italian Restaurant            Restaurant   Gym / Fitness Center   
5                    Pub              Wine Bar            Coffee Shop   
14  Gym / Fitness Center                 Plaza  Vietnamese Restaurant   
15  Gym / Fitness Center    Italian Restaurant           Cocktail Bar   
16            Restaurant  Gym / Fitness Center       Sushi Restaurant   
17                 Hotel        Sandwich Place     Italian Restaurant   
18  Gym / Fitness Center                 Hotel       Asian Restaurant   
19  Gym / Fitness Center              Wine Bar      French Restaurant   
20  Gym / Fitness Center                 Hotel             Restaurant   
21      Sushi Restaurant                 Plaza   Gym / Fitness Center   
22                  Café        Scenic Lookout            Art Gallery   
23  Gym / Fitness Center                  Café                  Hotel   

         6th Most Common Venue       7th Most Common Venue  \
0                          Pub          Seafood Restaurant   
1        Vietnamese Restaurant            Asian Restaurant   
2           Italian Restaurant                       Hotel   
3                Historic Site                       Hotel   
5           Italian Restaurant                       Plaza   
14              Scenic Lookout                Burger Joint   
15                       Hotel                  Steakhouse   
16           French Restaurant          Italian Restaurant   
17                 Art Gallery                         Pub   
18          Seafood Restaurant                Cocktail Bar   
19              Clothing Store            Asian Restaurant   
20                Cocktail Bar                  Steakhouse   
21  Modern European Restaurant                Cocktail Bar   
22              Clothing Store                  Steakhouse   
23              Clothing Store  Modern European Restaurant   

         8th Most Common Venue       9th Most Common Venue  \
0             Asian Restaurant  Modern European Restaurant   
1   Modern European Restaurant            Sushi Restaurant   
2                Historic Site        Fast Food Restaurant   
3                   Steakhouse            Asian Restaurant   
5                         Park                    Beer Bar   
14              Clothing Store  Modern European Restaurant   
15          English Restaurant                         Pub   
16                         Pub                Cocktail Bar   
17            Sushi Restaurant                  Food Truck   
18                        Café                  Steakhouse   
19                  Steakhouse                       Hotel   
20    

2nd Cluster

In [141]:
London_merged.loc[London_merged['Cluster Labels'] == 1, London_merged.columns[[0] + list(range(4, London_merged.shape[1]))]]

Ward 1st Most Common Venue 2nd Most Common Venue  \
6    Bishopsgate           Coffee Shop          Cocktail Bar   
7      Portsoken                 Hotel           Coffee Shop   
8    Lime Street           Coffee Shop          Cocktail Bar   
9      Langbourn                 Hotel  Gym / Fitness Center   
10       Aldgate                 Hotel  Gym / Fitness Center   
11  Billingsgate                 Hotel            Restaurant   
12        Bridge                 Hotel            Restaurant   
13         Tower                 Hotel           Coffee Shop   

   3rd Most Common Venue     4th Most Common Venue 5th Most Common Venue  \
6   Gym / Fitness Center         Indian Restaurant           Pizza Place   
7           Cocktail Bar      Gym / Fitness Center                  Café   
8                  Hotel  Mediterranean Restaurant           Pizza Place   
9            Coffee Shop         French Restaurant                   Pub   
10          Cocktail Bar                Restaurant           Coffee Shop   
11           Coffee Shop      Gym / Fitness Center                   Pub   
12           Coffee Shop                       Pub  Gym / Fitness Center   
13     Indian Restaurant      Gym / Fitness Center          Cocktail Bar   

       6th Most Common Venue 7th Most Common Venue  \
6   Mediterranean Restaurant           Salad Place   
7                        Pub     Indian Restaurant   
8                Salad Place            Restaurant   
9                Salad Place          Cocktail Bar   
10         French Restaurant                Garden   
11         French Restaurant                  Café   
12              Cocktail Bar         Historic Site   
13                       Pub                  Café   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
6                          Pub                 Hotel             Restaurant   
7                  Pizza Place           Salad Place     Turkish Restaurant   
8         Gym / Fitness Center    Italian Restaurant       Sushi Restaurant   
9                   Restaurant    English Restaurant     Italian Restaurant   
10          English Restaurant                   Pub            Salad Place   
11                Cocktail Bar           Salad Place       Asian Restaurant   
12                        Café           Salad Place       Asian Restaurant   
13  Modern European Restaurant           Salad Place                 Castle   

   11th Most Common Venue     12th Most Common Venue 13th Most Common Venue  \
6      Italian Restaurant        Japanese Restaurant             Boxing Gym   
7   Vietnamese Restaurant  Middle Eastern Restaurant       Asian Restaurant   
8          Breakfast Spot                 Boxing Gym                 Lounge   
9                  Garden              Historic Site     Falafel Restaurant   
10                   Café             Scenic Lookout              Hotel Bar   
11                 Garden         Turkish Restaurant   Fast Food Restaurant   
12                 Garden          French Restaurant     Italian Restaurant   
13              Hotel Bar          French Restaurant                 Garden   

   14th Most Common Venue     15th Most Common Venue  
6          Sandwich Place  Middle Eastern Restaurant  
7      English Restaurant                  Hotel Bar  
8            Burger Joint        Japanese Restaurant  
9                Wine Bar                       Café  
10      Indian Restaurant         Italian Restaurant  
11           Burger Joint                   Beer Bar  
12     English Restaurant               Burger Joint  
13            Pizza Place       Fast Food Restaurant

In [142]:
third=London_merged.loc[London_merged['Cluster Labels'] == 2, London_merged.columns[[0] + list(range(4, London_merged.shape[1]))]]
third

Ward 1st Most Common Venue 2nd Most Common Venue  \
4       Castle Baynard           Coffee Shop    Italian Restaurant   
24  Farringdon Without           Coffee Shop                   Pub   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4                    Pub        Sandwich Place     French Restaurant   
24        Sandwich Place     French Restaurant  Gym / Fitness Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4   Gym / Fitness Center              Wine Bar                 Hotel   
24    Italian Restaurant                 Hotel     Korean Restaurant   

   9th Most Common Venue 10th Most Common Venue 11th Most Common Venue  \
4          Burrito Place               Beer Bar   Fast Food Restaurant   
24                   Bar   Fast Food Restaurant    Japanese Restaurant   

   12th Most Common Venue 13th Most Common Venue 14th Most Common Venue  \
4             Salad Place    Japanese Restaurant                    Bar   
24             Restaurant                   Café              Bookstore   

   15th Most Common Venue  
4      Falafel Restaurant  
24         History Museum

In [143]:
first['Ward'].unique()

array(['Queenhithe', 'Vintry', 'Candlewick', 'Dowgate',
       'Farringdon Within', 'Bread Street', 'Cornhill', 'Broad Street',
       'Cripplegate', 'Walbrook', 'Cordwainer', 'Coleman Street',
       'Aldersgate', 'Bassishaw', 'Cheap'], dtype=object)

From the exploration of the clusters I could say that the first cluster (number zero) is the optimal one, since in it we can see a much higher concentration of Italian restaurants and, in a few cases, pizza places between the first 15 venues. 

To choose the best Wards I based my decision on the concentartion of competition. In this specific case I want to precise that the competition I looked for is the one made from Italian restaurants because of the nature of the pizza place that I want to open.

Best Wards where in my opinion is convinient to open a pizza place are:
- Vintry 
- Bread Street
- Queenhithe 
- Walbrook
- Cordwainer
- Coleman Street
- Aldersgate
- Bassishaw
- Cheap